DBZ-1928 updated doc for Avro serialization

TovaCohen opened the pull request:
Here's the updated doc for Avro serialization. There is lots of conditionalization. Please review this carefully.

Naros commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -9,57 +13,92 @@
-== The Apicurio API and Schema Registry
+* For an individual connector. 
++
+Kafka Connect comes with a _JSON converter_ that serializes message keys and values into JSON documents. You can configure the JSON converter to include or exclude the message schema by specifying the `key.converter.schemas.enable` and `value.converter.schemas.enable` properties. The {prodname} {link-prefix}:{link-tutorial}[tutorial] shows what the messages look like when both payload and schemas ...
```
I feel like this and the following two parameters apply to both Kafka worker or individual connector setups and probably shouldn't be indented under the second option.

TovaCohen replied:
Sure. I removed the plus signs that indent those paragraphs. Fixed in the next commit.

Naros commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -199,3 +358,5 @@ Note that some details around Kafka Connect converters have slightly changed sin
 For a complete example of using Avro as the message format for {prodname} data change events,
 please see the https://github.com/debezium/debezium-examples/tree/master/tutorial#using-mysql-and-the-avro-message-format[MySQL and the Avro message format] tutorial example.
+
+endif::[community]
```
Looks like this is a typo, should be `endif::community[]`?

TovaCohen replied:
Yes, that's a typo. Fixed in the next commit.

TovaCohen commented:
Sorry about the third commit - I forgot to remove the temporary :community: declaration that I needed to confirm that the content renders correctly.

gunnarmorling commented:
Test failure is unrelated and tracked via https://issues.redhat.com/browse/DBZ-2231.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -9,57 +13,92 @@
 
-You can install them into any Kafka distribution and use them with Kafka Connect.
+Alternatively, you can serialize the message keys and values by using link:https://avro.apache.org/[Apache Avro].
+The Avro binary format is compact and efficient, and Avro schemas make it possible to ensure that the messages have the correct structure. Avro's schema evolution mechanism makes it possible to evolve the schemas over time, which is essential for {prodname} connectors that dynamically generate the message schemas to match the structure of the database tables. Over time, the change events captured ...
```
Without the section before (which now is part of the Apicurio section), there's some context lacking _why_ a registry is needed. How about adding something like this here:

When using Apache Avro as a message format, a schema registry must be deployed which manages Avro message schemas and their versions. Options include the Apicurio API and Schema Registry and the Confluent Schema registry.

TovaCohen replied:
Before this point in the source file, no content is conditionalzed for just upstream nor just downstream. That is, all content is visible both upstream and downstream. But I think you are right that the doc needs to say that a registry is required and why. I added the following at the end of the section: 

To use Apache Avro as the message format, you must deploy a schema registry that manages Avro message schemas and their versions.

I left out you suggested second sentence because there is only one option downstream, and the next section in the source file describes it. Then a bit later, conditionalized for upstream only, the doc says you can use Confluent.

gunnarmorling replied:
Sounds good.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -9,57 +13,92 @@
+====
+
+ifdef::community[]
+Another option is using the Confluent schema registry, which is described later.
```
A reference would be nice.

TovaCohen replied:
Sure. I added one.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -9,57 +13,92 @@
+. Install the Avro converter by downloading the {prodname} link:https://access.redhat.com/jbossnetwork/restricted/listSoftware.html?product=red.hat.integration&downloadType=distributions[Service Registry Kafka Connect] zip file and extracting it into the {prodname} connector's directory.
+endif::product[]
+
+. Configure a Kafka Connect instance with the following property settings: 
```
Can you add a sentence after the listing like this: Alternatively, you can also configure specific connectors to use the Avro converter.

TovaCohen replied:
Can the word "also" be removed from the sentence that you want to add?
And should this additional sentence be indented to align with the listing? I think so, but I'm not certain.

gunnarmorling replied:
Yes, you can remove the "also". Re indentiation, I don't know, perhaps even just in brackets after "property settings (alternatively...)".

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -69,30 +108,81 @@ value.converter.apicurio.registry.url=http://apicurio:8080/api
+// ModuleID: deploying-connectors-that-use-avro-in-debezium-containers
+== Deploying with {prodname} containers
+
+In your environment, you might want to use a {prodname} pre-built container to deploy a {prodname} connector that uses Avro serializaion. Follow the procedure here to do that. In this procedure, you build a {prodname} image in which {prodname} uses the Avro converter.  
```
```suggestion
In your environment, you might want to use a {prodname} pre-built container to deploy {prodname} connectors that use Avro serialization. Follow the procedure here to do that. In this procedure, you build a custom Kafka Connect container image for {prodname} which contains the Avro converter.  
```

TovaCohen replied:
Done.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -69,30 +108,81 @@ value.converter.apicurio.registry.url=http://apicurio:8080/api
+The in-memory deployment is not suitable for production. Use the Kafka Streams persistence option for production. For more information, see {LinkServiceRegistryGetStart}[NameServiceRegistryGetStart].
+endif::product[]
+
+. Build a {prodname} container image that implements the Avro converter:
```
```suggestion
. Build a {prodname} container image that contains the Avro converter:
```

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -69,30 +108,81 @@ value.converter.apicurio.registry.url=http://apicurio:8080/api
+ARG DEBEZIUM_VERSION
+FROM debezium/connect:$DEBEZIUM_VERSION
+ENV KAFKA_CONNECT_DEBEZIUM_DIR=$KAFKA_CONNECT_PLUGINS_DIR/debezium-connector-mysql
+ENV APICURIO_VERSION=1.1.2.Final
```
There's a variable defined for the version (in antora.yml in case of upstream docs): {apicurio-version}. Enable substitution for the listing in order to reference it.

TovaCohen replied:
Done. In next commit.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -69,30 +108,81 @@ value.converter.apicurio.registry.url=http://apicurio:8080/api
+.. Create a custom image for Kafka Connect. See link:{LinkCDCInstallOpenShift}[NameCDCInstallOpenShift] for an example of how to do this. Start with the `Dockerfile` in that example. Then add the {registry} converters to the connector directories. 
+endif::product[]
+
+. Run a Kafka Connect image that is configured to use Avro:
```
```suggestion
. Run the newly built Kafka Connect image, configuring it so it uses the Avro converter (or configure specific converters to do so):
```

TovaCohen replied:
I don't understand the last part of your suggestion. 
What do you mean by "configure specific converters to do so"? -- It sounds like this says that you can configure other converters to use the Avro converter. Is that what you mean?

gunnarmorling replied:
Ah, sorry, I meant "specific connectors". Converters can be configured globally for the entire Connect worker or individually for connectors. As commented somewhere else, I think ideally we'd only discuss the latter.

TovaCohen replied:
Fixed for now. Will revisit to do better.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -69,30 +108,81 @@ value.converter.apicurio.registry.url=http://apicurio:8080/api
+.. Choose the OpenShift project in which you want to deploy the {prodname} connector. In the following command, `$NAMESPACE` represents your project.  
+.. Deploy the latest published {registry} operator by running the following command:
++
+`curl -sSL https://raw.githubusercontent.com/apicurio/apicurio-registry-operator/master/docs/resources/install.yaml | sed "s/{NAMESPACE}/$NAMESPACE/g" | oc apply -f -`
```
Is this the right resource for downstream?

TovaCohen replied:
I asked Steve McCarthy, to point me to the downstream resource and this is what he pointed me to. I don't know his GitHub user name. I'll find out and tag him here to confirm.

gunnarmorling replied:
Ok, let's clarify that. In particular using "master" seems really odd for product, I'd expect a fixed reference instead of a live branch.

TovaCohen replied:
Adding @smccarthy-ie , to help us straighten this out.

smccarthy-ie replied:
@TovaCohen @gunnarmorling, OK, some confusion around upstream/downstream product names. You were asking about Apicurio Registry (upstream) installation, but Service Registry is the downstream product :)  

The latest downstream Service Registry docs show using the OperatorHub in the OpenShift UI. I don't know the command line equivalent. 

@jsenko Can you tell us what is the downstream equivalent of `curl -sSL https://raw.githubusercontent.com/apicurio/apicurio-registry-operator/master/docs/resources/install.yaml | sed "s/{NAMESPACE}/$NAMESPACE/g" | oc apply -f -`

TovaCohen replied:
I replaced these instructions with a cross-reference to follow the instructions in Getting Started with Service Registry, "Installing Service Registry from the OperatorHub".
@gunnarmorling  Let me know if you think we also need to provide a command line alternative.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -69,30 +108,81 @@ value.converter.apicurio.registry.url=http://apicurio:8080/api
++
+`curl -sSL https://raw.githubusercontent.com/apicurio/apicurio-registry-operator/master/docs/resources/install.yaml | sed "s/{NAMESPACE}/$NAMESPACE/g" | oc apply -f -`
++
+This deploys the latest development version of the {registry} operator from the `master` branch. To deploy other versions, specify a different branch or tag, or edit the operator image reference in the file.
```
This surely should be a stable released version, not master?

TovaCohen replied:
I will ask Steve McCarthy to confirm.

TovaCohen replied:
Adding @smccarthy-ie , to help us straighten this out.

jsenko replied:
Latest stable released version is `0.0.2` (tag name is the same)

TovaCohen replied:
@gunnarmorling  I removed this instruction. Let me know if you want me to put back an updated command line.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -69,30 +108,81 @@ value.converter.apicurio.registry.url=http://apicurio:8080/api
++
+`oc create -f https://raw.githubusercontent.com/apicurio/apicurio-registry-operator/master/docs/resources/example-cr/in-memory.yaml`
++
+The in-memory deployment is not suitable for production. Use the Kafka Streams persistence option for production. For more information, see {LinkServiceRegistryGetStart}[NameServiceRegistryGetStart].
```
```suggestion
The in-memory deployment is not suitable for production. Use the Apache Kafka persistence option for production. For more information, see {LinkServiceRegistryGetStart}[NameServiceRegistryGetStart].
```

TovaCohen replied:
Done

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -112,19 +202,90 @@ docker run -it --rm --name connect \
+      database.whitelist: inventory  
+      database.history.kafka.bootstrap.servers: my-cluster-kafka-bootstrap:9092  
+      database.history.kafka.topic: schema-changes.inventory  
+      key.converter: io.apicurio.registry.utils.converter.AvroConverter
+      key.converter.apicurio.registry.url: http://apicurio:8080/api
+      key.converter.apicurio.registry.global-id: io.apicurio.registry.utils.serde.strategy.AutoRegisterIdStrategy
+      value.converter: io.apicurio.registry.utils.converter.AvroConverter
+      value.converter.apicurio.registry.url: http://apicurio:8080/api
+      value.converter.apicurio.registry.global-id: io.apicurio.registry.utils.serde.strategy.AutoRegisterIdStrategy
```
Ok, so that's good to show this connector-specific configuration here. Taking a step back, there's still some inconsistency overall. The situation is this:

* Avro can be enabled on the Kafka Connect worker level (as shown above via "Configure a Kafka Connect instance with the following property settings", and also then via the environment variables in the community variant)
* Avro can be enabled for a specific connector (as shown here for product, it's currently not shown for the community variant)

My general recommendation would be to _only_ show the second approach, enabling it for a specific connector, and not mention the worker-level approach, as it's less flexible.

TovaCohen replied:
I don't know how to implement your recommendation - that is, I don't know how to change the doc. I'll commit the changes I made and then perhaps you can suggest the specific changes you want to see. I'll be happy to accept/commit your suggestions. This time around, I didn't Commit Suggestions because in the past Chris has recommended one commit with all the updates.

gunnarmorling replied:
Ok. Let me know once you're done from your side and I'll try and make that change myself.

gunnarmorling replied:
In general though, the idea is to solely show how the Avro converter is used at the connector level instead of showing how to enable it globally for the Kafka Connect worker node. That's already the case for the product variant. For upstream it'd mean to show the right curl POST command with the equivalent configuration.

TovaCohen replied:
I removed all references to workers. The doc now talks only about configuring a Debezium connector, upstream and downstream.

TovaCohen commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -9,57 +13,92 @@
+
+// Type: concept
+// Title: Overview of deploying a {prodname} connector that uses Avro serialization
+[id="overview-of-deploying-a-debezium-connector-that-uses-avro-serializatioin"]
```
Just noting here that I fixed this in the source. I will commit all the updates in one operation.

TovaCohen commented:
I did all the updates that I can at this point. Please feel free to make and commit additional updates.

TovaCohen commented:
Just noting here that I am going to try to understand how to update the doc for configuring workers vs individual connectors. Need to form my questions so that I can do a good job. Monday.

gunnarmorling commented:
> Just noting here that I am going to try to understand how to update the doc for configuring workers vs individual connectors

👍  Here's resource that could be helpful for that: https://github.com/debezium/debezium-examples/tree/master/tutorial#using-mysql-and-the-avro-message-format. There we show both flavours, but for upstream (i.e. Kafka and Connect started via Docker, connectors registered via curl). That should help for the upstream part, and for downstream the CR-based approach is already part of this PR.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -9,57 +13,94 @@
+* For an individual connector. 
 
-The open-source project https://github.com/Apicurio/apicurio-registry[Apicurio Registry] provides several components that work with Avro:
+Kafka Connect comes with a _JSON converter_ that serializes message keys and values into JSON documents. You can configure the JSON converter to include or exclude the message schema by specifying the `key.converter.schemas.enable` and `value.converter.schemas.enable` properties. The {prodname} {link-prefix}:{link-tutorial}[tutorial] shows what the messages look like when both payload and schemas ...
```
I don't think the tutorial link will work downstream, or would it? Rather link to the "Getting Started" documented for that?

TovaCohen replied:
At this point, I cannot publish a rendered version of this content on the Pantheon Staging site. The Avro content is hidden for the June 25th release. On June 25th, we publish the Stage version of the Debezium User Guide. So it cannot show the Avro serialization content.  I hope that's clear. Let me know if not!

gunnarmorling replied:
Sorry, I'm not quite sure I'm following. I was referring to the link `{prodname} {link-prefix}:{link-tutorial}[tutorial]` which points to a resource (tutorial) which doesn't exist in downstream docs. Instead, the "Getting Started" guide should be linked in a conditional for downstream.

TovaCohen replied:
As for your question about the link to the tutorial - yes, that will work downstream. The downstream versions of these attributes are set to display the customer portal Debezium Getting Started Guide. I think that the clickable text of "tutorial" is okay as a generic term. But if you prefer to see "Debezium Getting Started Guide" as the clickable text, let me know. I would need to define a new document attribute to do that.

Naros commented:
Hi @TovaCohen, where are we on this PR?  Looks like we have some pending questions and I'm not clear whether those have been addressed based on the lack of recent activity.

TovaCohen commented:
This is still a work in progress. 
Since we are hiding this doc downstream until the June 30 release of Service Registry, I have higher priority tasks right now. 
If you want to merge this as it is,  I think that's okay because it is the downstream-only content that needs work.  But you probably should confirm that.

gunnarmorling commented:
> But if you prefer to see "Debezium Getting Started Guide" as the
clickable text...

Aaah, gotcha. Personally I think "tutorial" is a tad confusing downstream,
as we don't use that term elsewhere there, but I'll leave it to you. Thanks!

>

TovaCohen commented:
@gunnarmorling  So the downstream version of the instructions for running the newly built Kafka Connect image refers to the S2I image and has cross-references to the Getting Started guide, which of course still shows how to use the S2I image to deploy the connector. Should I leave these references to S2I in place? Or should we update the instructions to use the Kafka Connect image in the container catalog, in the same way that we updated Installing Debezium on OpenShift, Creating a container image from the Kafka Connect base image. ??

TovaCohen commented:
In the updates that I just pushed, I removed the references to Kafka Connect workers. 
This downstream instructions for deploying a Debezium connector that uses Avro still needs updates. There is a disconnect/inconsistency in those instructions. In step 2, I point to “Installing Debezium on OpenShift”, Creating a container from the Kafka Connect base image, which we just updated. But then in step 3, I point to the Debezium Getting Started Guide, example of deploying Kafka Connect, which uses the S2I image example. I don’t know how to update step 3 to correctly follow step 2. 
I added an attribute for the name of the tutorial. Upstream, the value is "tutorial". Downstream, the value will be "Debezium Getting Started Guide". 
There is also an unrelated update in cloudevents.adoc. I had to insert a plus sign in a source listiing heading so that antora would correctly build the content. I think I removed that plus sign elsewhere. But now I understand that it is needed to correctly substitute attribute values in source listings. 
Thanks in advance for your help in getting this right!

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -112,26 +194,98 @@ docker run -it --rm --name connect \
+endif::community[]
+ifdef::product[]
+
+.. Follow the steps in the link:{LinkCDCGettingStarted}#deploying-kafka-connect[example of deploying Kafka Connect] in {NameCDCGettingStarted}.
```
@TovaCohen, I somehow had missed this before (thanks to @jcechace for pointing it out), here you're adding new S2I docs. While we agreed to only fully transition to the custom containers approach for the next release docs, why adding more S2I-flavoured this time, which we only will have to modify next time?

TovaCohen replied:
@gunnarmorling I agree that this is the wrong way to update the doc. But I don't know how to make it right. Please see my previous comment.  I recognize the inconsistency and the disconnect in the procedure. But I don't have any background with regard to OpenShift images. I plan to remedy that. I'll go do some reading now. But I need precise input from you or Jiri or someone to correct this documentation for this release.

jcechace replied:
Not only that... this procedure starts with Dockerfile (correct way of building custom Kafka Connect image) and then switches to s2i... this creates a mix between two working (although we want to promote only the Docker way) approaches...

jcechace commented:
@TovaCohen  do we have the docs for Registry getting started? The link in these docs doesn't seem to work.

TovaCohen commented:
@jcechace The RH Service Registry documents have not been published, so the links in the Debezium Avro content won't work yet. If you want to see the content in the Service Registry documents, you can look at the Stage versions here: 
https://pantheon.int.us-west.aws.prod.paas.redhat.com/#/titles/red_hat_integration
And this is the link for the Stage version of Getting Started with Service Registry:
https://access.redhat.com/documentation/en-us/red_hat_integration/2020-q2/html-single/getting_started_with_service_registry/index?lb_target=stage

jcechace commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -9,57 +13,94 @@
+. Deploy an link:https://github.com/Apicurio/apicurio-registry[{registry-name-full}] instance.
+endif::community[]
+ifdef::product[]
+. Deploy a link:{LinkServiceRegistryGetStart}[{registry-name-full} instance by following the instructions in {NameServiceRegistryGetStart}].
```
This doesn't render (yet?) into working link

TovaCohen replied:
Right. This doc has not yet been published. It should not work yet.

jcechace commented on documentation/modules/ROOT/pages/configuration/avro.adoc:
```diff
@@ -69,30 +110,71 @@ value.converter.apicurio.registry.url=http://apicurio:8080/api
+.. Download the link:https://access.redhat.com/jbossnetwork/restricted/listSoftware.html?product=red.hat.integration&downloadType=distributions[{registry} Kafka Connect] zip file. 
+.. Extract the content into the directory that contains the {prodname} connector that you are configuring to use Avro serialization. 
+.. Create a custom image for Kafka Connect. See link:{LinkCDCInstallOpenShift}[{NameCDCInstallOpenShift}, Creating a container image from the Kafka Connect base image] for an example of how to do this. Start with the `Dockerfile` in that example. Then add the {registry} converters to the connector directories. 
+endif::product[]
```
This section talks about Docker file, which means building a custom Kafka Connect image (which is correct), the issue comes in my next comment

TovaCohen commented:
I updated the downstream version of the procedure for configuring and deploying Debezium connectors that are configured to use Avro serialization. This update needs a careful review. 
To help you review it, I manually copied the latest avro.adoc file into my downstream repo and created a WIP merge request. So you can see a preview of the downstream procedure here: 
https://cee-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/CCS/job/ccs-mr-preview/12567/artifact/docs/change-data-capture-user-guide/preview/index.html#avro-serialization
Let me know what needs to change.

TovaCohen commented:
@jcechace @gunnarmorling @jpechane @Naros 
I  manually copied the updated avro.adoc file to my downstream repo, confirmed that the Debezium User Guide builds, committed, pushed and merged the update. You can see the rendered downstream version of the Avro content here:
https://access.redhat.com/documentation/en-us/red_hat_integration/2020-q2/html-single/debezium_user_guide/index?lb_target=stage#avro-serialization
I did this to make it easier for you to review. Let me know what needs to change. 
When this PR gets merged and backported to 1.1, then in my downstream repo, I will fetch the updates from the 1.1 branch. 
Thanks!

jcechace commented:
LGTM

Naros commented:
Applied to master & back-ported to 1.1.
